src/mono/mono/arch/amd64/amd64-codegen.h

```diff
@@ -1216,9 +1216,25 @@ typedef union {
 
 #define amd64_prefix_size(inst,p,size) do { x86_prefix((inst), p); } while (0)
 #define amd64_rdtsc_size(inst,size) do { amd64_codegen_pre(inst); amd64_emit_rex ((inst),(size),0,0,0); x86_rdtsc(inst); amd64_codegen_post(inst); } while (0)
+// FIXME: wrong for size == 1 or 2
 #define amd64_cmpxchg_reg_reg_size(inst,dreg,reg,size) do { amd64_codegen_pre(inst); amd64_emit_rex ((inst),(size),(dreg),0,(reg)); x86_cmpxchg_reg_reg((inst),((dreg)&0x7),((reg)&0x7)); amd64_codegen_post(inst); } while (0)
+// FIXME: wrong for size == 1 or 2
 #define amd64_cmpxchg_mem_reg_size(inst,mem,reg,size) do { amd64_codegen_pre(inst); amd64_emit_rex ((inst),(size),0,0,(reg)); x86_cmpxchg_mem_reg((inst),(mem),((reg)&0x7)); amd64_codegen_post(inst); } while (0)
-#define amd64_cmpxchg_membase_reg_size(inst,basereg,disp,reg,size) do { amd64_codegen_pre(inst); amd64_emit_rex ((inst),(size),(reg),0,(basereg)); x86_cmpxchg_membase_reg((inst),((basereg)&0x7),(disp),((reg)&0x7)); amd64_codegen_post(inst); } while (0)
+
+#define amd64_cmpxchg_membase_reg_size(inst,basereg,disp,reg,size) \
+	do { \
+		amd64_codegen_pre(inst); \
+		if ((size) == 2) \
+			x86_prefix((inst), X86_OPERAND_PREFIX); \
+		amd64_emit_rex ((inst),(size),(reg),0,(basereg)); \
+		switch ((size)) { \
+		case 1: x86_byte((inst), 0x0f); x86_byte((inst), 0xb0); break; \
+		case 2: case 4: case 8: x86_byte((inst), 0x0f); x86_byte((inst), 0xb1); break; \
+		default: assert (0); \
+		}\
+		x86_membase_emit((inst),((reg)&0x7),((basereg)&0x7),(disp)); \
+		amd64_codegen_post(inst); \
+	} while (0)
```

Comment:
@JulieLeeMSFT can someone from the JIT team please check that I'm doing x86-64 instruction encoding correctly here.

Things I'm unclear about:
1. can we really emit a REX prefix for 16-bit operands?  I'm not sure what in mono's codegen will guarantee that `reg` and `basereg` are the legacy registers if REX isn't allowed.
2. does this work if `reg` or `basereg` is one of the R8-R15 registers and size is 1? (apparently it worked when size was 4 or 8)

Reply:
Answering my own question: https://godbolt.org/z/jPEzErvGK

1. looks like `X86_OPERAND_PREFIX` and REX can co-exist.
2. seems to work if `reg` is one of the new registers and size is 1